### PR TITLE
feat: routing

### PR DESCRIPTION
**Pull request by codeflyer**

### CODE REVIEW

File: api/basic/add-pin-router.ts [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    TABLE_NAME: table = '',
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    // set for testing
+    DYNAMO_DB_ENDPOINT: dbEndpoint = undefined,
+    INDEXER_ENDPOINT: indexerEndpoint = '',
+    PICKUP_ENDPOINT: pickupEndpoint = '',
+    BALANCER_RATE: balancerRate = 100
+  } = process.env
+
+  const authResponse = doAuth(event.headers.authorization)
```
Comment (simone-sanfratello) [inline api/basic/add-pin-router.ts:41]: looks like `authError` instead of `authResponse`

File: api/basic/add-pin-router.ts [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+  if (authResponse != null) return authResponse
+
+  const dynamo = new DynamoDBClient({ endpoint: dbEndpoint })
+  const cid = event.pathParameters?.cid ?? ''
+  const origins = event.queryStringParameters?.origins?.split(',') ?? []
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!table) {
```
Comment (simone-sanfratello) [inline api/basic/add-pin-router.ts:49]: This control can be the first one, before checking auth and create the dynamo client, to fail fast

File: api/basic/add-pin-router.ts [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+    return {
+      statusCode: 500,
+      body: JSON.stringify({ error: { reason: 'INTERNAL_SERVER_ERROR', details: 'TABLE must be set in ENV' } })
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!cid) {
```
Comment (simone-sanfratello) [inline api/basic/add-pin-router.ts:57]: same, those validations can be the fist ones and fail fast

File: api/basic/add-pin-router.ts [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+  table,
+  indexerEndpoint,
+  pickupEndpoint,
+  token,
+  balancerRate
+}: AddPinInput): Promise<Response> {
+  const pinFromDynamo = await getPinFromDynamo(dynamo, table, cid)
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!pinFromDynamo) {
```
Comment (simone-sanfratello) [inline api/basic/add-pin-router.ts:117]: The flow is a little bit hard to follow, I'd use early returns (not necessary)

File: api/basic/get-pin-router.ts [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+ * see: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format
+ */
+export async function handler (event: APIGatewayProxyEventV2): Promise<Response> {
+  const {
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    INDEXER_ENDPOINT: indexerEndpoint = '',
+    PICKUP_ENDPOINT: pickupEndpoint = ''
+  } = process.env
+
+  const authResponse = doAuth(event.headers.authorization)
```
Comment (simone-sanfratello) [inline api/basic/get-pin-router.ts:22]: probably `authError`

File: api/basic/get-pin-router.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+  if (authResponse != null) return authResponse
+
+  const cid = event.pathParameters?.cid ?? ''
+
+  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
+  if (!cid) {
```
Comment (simone-sanfratello) [inline api/basic/get-pin-router.ts:28]: I'd do the validations at first and fail fast

File: api/test/use-pickup.test.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  const NUM_TEST = 100
+  const RATE = 5
+  for (let i = 0; i < NUM_TEST * 100; i++) {
+    if (usePickup(RATE)) {
+      isTrue++
+    }
+  }
+
+  // The function works with a random number, it's tested using range of error of the 30%
+  t.true(isTrue / NUM_TEST > RATE * 0.70)
```
Comment (simone-sanfratello) [inline api/test/use-pickup.test.js:15]: I'd name the constant as TOLLERANCE or so, it's not clear what they do/are

File: api/test/use-pickup.test.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
+import test from 'ava'
+import { usePickup } from '../basic/add-pin-router.js'
+
+test('usePickup with rate of 5', async t => {
```
Comment (simone-sanfratello) [inline api/test/use-pickup.test.js:4]: Consider an helper function to run with different params

**Review by simone-sanfratello — CHANGES_REQUESTED** [review]
Move validations on top of handlers
Other comments are optional

**Review by simone-sanfratello — APPROVED** [review]

File: api/basic/add-pin-router.ts [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+ * We provide responses in Payload format v2.0
+ * see: https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format
+ */
+export async function handler (event: APIGatewayProxyEventV2): Promise<Response> {
+  const {
+    TABLE_NAME: table = '',
+    CLUSTER_BASIC_AUTH_TOKEN: token = '',
+    // set for testing
+    DYNAMO_DB_ENDPOINT: dbEndpoint = undefined,
+    INDEXER_ENDPOINT: indexerEndpoint = '',
```
Comment (francardoso93) [inline api/basic/add-pin-router.ts:36]: Misconception: There are no indexers involved in this process. What you're calling indexer is actually the IPFS Cluster. See my other comment on CLUSTER_IPFS_ADDR

File: stacks/BasicApiStack.ts [hunk 10]
```diff
@@ -51,13 +51,18 @@ export function BasicApiStack ({ app, stack }: StackContext): { queue: Queue, bu
           TABLE_NAME: table.tableName,
           QUEUE_URL: queue.queueUrl,
           CLUSTER_BASIC_AUTH_TOKEN: process.env.CLUSTER_BASIC_AUTH_TOKEN ?? '',
-          CLUSTER_IPFS_ADDR: process.env.CLUSTER_IPFS_ADDR ?? ''
+          CLUSTER_IPFS_ADDR: process.env.CLUSTER_IPFS_ADDR ?? '',
```
Comment (francardoso93) [inline stacks/BasicApiStack.ts:54]: Misconception: What you're setting here is actually the address of ELASTIC IPFS, and to CLUSTER IPFS.
Elastic IPFS is our product running in the cloud.
Cluster IPFS is an on prem setup where they currently store pinned blocks.

File: api/basic/add-pin-router.ts (hunk 9 above)
Comment (francardoso93) [inline api/basic/add-pin-router.ts:36]: Misconception: There are no indexers involved in this process. What you're calling indexer is actually the IPFS Cluster. See my other comment on CLUSTER_IPFS_ADDR

File: stacks/BasicApiStack.ts (hunk 10 above)
Comment (francardoso93) [inline stacks/BasicApiStack.ts:54]: Misconception: What you're setting here is actually the address of `Elastic IPFS`, and not `Cluster IPFS`.
- Elastic IPFS is our product running in the cloud. 
- Cluster IPFS is an on prem setup where they currently store pinned blocks.

**Review by francardoso93 — CHANGES_REQUESTED** [review]

File: api/basic/add-pin-router.ts (hunk 2 above)
Comment (codeflyer) [reply api/basic/add-pin-router.ts:49]: The auth should be the first, if the user is not authorised should not see any specific error.

Then I'll group configuration and parameters errors

File: api/basic/add-pin-router.ts (hunk 4 above)
Comment (codeflyer) [reply api/basic/add-pin-router.ts]: cleaned up

File: api/basic/get-pin-router.ts (hunk 6 above)
Comment (codeflyer) [reply api/basic/get-pin-router.ts]: I've cleaned up the validation creating some helpers and given precedence to the readability.

File: stacks/BasicApiStack.ts (hunk 10 above)
Comment (codeflyer) [reply stacks/BasicApiStack.ts:54]: Fixed as discussed

**Review by francardoso93 — APPROVED** [review]
I still see lot's of the word "indexer" in comments or other variable names
